### PR TITLE
Dynamic fixes

### PR DESCRIPTION
**Pull request by blinktag**

### CODE REVIEW

File: bus/broadcast_consumer_connection.go [hunk 1]
```diff
@@ -52,6 +53,26 @@ func (b *Bus) doUpdateConnection(_ context.Context, msg *Message) error {
 }
 
 func (b *Bus) doDeleteConnection(ctx context.Context, msg *Message) error {
+	// Ensure this connection isn't being used by any dynamic replays
```
Comment (dselans) [inline bus/broadcast_consumer_connection.go:56]: We should update the gRPC delete handler to have the same logic. I like this approach a lot better.

**Review by dselans — APPROVED** [review]

File: bus/broadcast_consumer_connection.go [hunk 2]
```diff
@@ -52,6 +53,26 @@ func (b *Bus) doUpdateConnection(_ context.Context, msg *Message) error {
...
+				id, dynamicReplay.Options.XDynamicId)
+		}
+	}
+	b.config.PersistentConfig.DynamicReplaysMutex.RUnlock()
+
+	// Ensure this connection isn't being used by any relays
+	b.config.PersistentConfig.RelaysMutex.RLock()
+	for id, dynamicReplay := range b.config.PersistentConfig.Relays {
+		if dynamicReplay.Options.ConnectionId == id {
+			return fmt.Errorf("cannot delete connection '%s' because it is in use by relay '%s'",
```
Comment (dselans) [inline bus/broadcast_consumer_connection.go:70]: Need to unlock here

File: bus/broadcast_consumer_connection.go [hunk 3]
```diff
@@ -52,6 +53,26 @@ func (b *Bus) doUpdateConnection(_ context.Context, msg *Message) error {
 }
 
 func (b *Bus) doDeleteConnection(ctx context.Context, msg *Message) error {
+	// Ensure this connection isn't being used by any dynamic replays
+	b.config.PersistentConfig.DynamicReplaysMutex.RLock()
+	for id, dynamicReplay := range b.config.PersistentConfig.Dynamic {
+		if dynamicReplay.Options.ConnectionId == id {
+			return fmt.Errorf("cannot delete connection '%s' because it is in use by dynamic replay '%s'",
```
Comment (dselans) [inline bus/broadcast_consumer_connection.go:60]: Need to unlock here

File: bus/broadcast_consumer_connection.go (hunk 2 above)
Comment (dselans) [reply bus/broadcast_consumer_connection.go:70]: I would honestly just do a lock at the beginning (or a lock + copy) and defer unlock.

File: dynamic/dynamic.go [hunk 4]
```diff
@@ -255,12 +257,19 @@ func (d *Client) connect(ctx context.Context, bus string) (services.DProxy_Conne
...
+				MessageBus:       bus,
+				PlumberClusterId: d.PlumberClusterID,
+				Name:             d.Options.Name,
 			},
 		},
 	}
 
+	// dProxy needs a unique ID to track the connection
+	// This value can be anything, so long as it is unique
+	md := metadata.Pairs("plumber-stream-id", uuid.NewV4().String())
```
Comment (dselans) [inline dynamic/dynamic.go:270]: Is it okay that it's going to be a diff plumber-stream-id every time? Is that the goal?

**Review by dselans — APPROVED** [review]
2 important things - missing mutex unlocks. Rest are just comments/curiosities.

File: dynamic/dynamic.go (hunk 4 above)
Comment (blinktag) [reply dynamic/dynamic.go:270]: Yes, it just needs to track it for the lifespan of the `Connect()` gRPC method https://github.com/batchcorp/dproxy/blob/e94c972fefed632ba990eef644f28b43fe394c9d/services/plumberproxy/grpc.go#L85-L88

File: bus/broadcast_consumer_connection.go (hunk 1 above)
Comment (blinktag) [reply bus/broadcast_consumer_connection.go:56]: Totally missed that, done!

File: bus/broadcast_consumer_connection.go (hunk 3 above)
Comment (blinktag) [reply bus/broadcast_consumer_connection.go:64]: Fixed!

**Pull request merged**